### ISSUE TITLE
Package request: ALVR

### ISSUE THREAD

**Issue by mlkood** [NixOS/nixpkgs#238102]
ALVR - Air Light VR

Stream VR games from your PC to your headset via Wi-Fi.

**Metadata**

* homepage URL: https://github.com/alvr-org/ALVR
* source URL: https://github.com/alvr-org/ALVR
* license: mit
* platforms: linux

**Comment by CaptainSolidus** [comment]
I'd also be interested in this being available as a nix package.

**Comment by Krutonium** [comment]
I would also like this.

**Comment by CJendantix** [comment]
building from source is a pain, PACKAGE THIS!!!

**Comment by czadikem** [comment]
Is this gonna happen?

**Comment by AndrewKvalheim** [comment]
Related: alvr-org/ALVR#1738

**Comment by NovaViper** [comment]
@PassiveLemon Hey I just wanted to comment and say thank you! I got ALVR to work and also got my games running with it too! I actually already had the appimage of ALVR's nightly version working on my NixOS install so it only a matter of just moving over my settings to the version being packaged. The firewall rules work also, which allows ALVR to automatically find my headset without having to add in the IP address which is a nice bonus! Will be uploading a commit onto [my personal nix-config](https://github.com/NovaViper/NixConfig) for reference

**Comment by PassiveLemon** [comment]
Thanks for testing and I'm glad it works for you!

**Comment by GHOSCHT** [comment]
Streaming with NVENC works really good for me, but sadly I'm really struggling with audio setup. It would be nice to know @NovaViper if you or anyone else had any similar problems, since my [nix-config](https://git.ghoscht.com/ghoscht/nix-config) is really similar to yours (Misterio77 configs). The ALVR setup is also based on your examples. ALVR is running with its stock config file.

ALVR somehow cannot connect to the pipewire backend, which means no pipewire node appears which I can route audio to.
I tried both ALSA and JACK backends, with the pipewire ALSA and JACK plugins for pipewire enabled ([pipewire.nix](https://git.ghoscht.com/ghoscht/nix-config/src/branch/main/hosts/common/optional/desktop/pipewire.nix))

When setting the audio device to "pipewire" in ALSA mode, which according to the wiki is recommended I get the following error:
`Handshake error for 8576.client.alvr: Cannot find audio device which name contains "pipewire"`
Which is really weird, since I can successfully play audio via the "pipewire" ALSA device using `speaker-test -D pipewire` from `pkgs.alsa-utils`. Using that test command a new node in qtpwgraph appears, so I assume that the underlying pipewire setup works.

What's also strange is, that I am able to select the "default" ALSA audio device with no errors. While I can hear test sounds using `speaker-test -D default` and see speaker-test create a new pipewire node, ALVR simply doesn't do that.

A similar error appears when I choose the JACK backend, with "cpal_client_out" and "cpal_client_in":
`Handshake error for 8576.client.alvr: Cannot find audio device which name contains "cpal_client_out"`

I can however access existing ALSA devices, which is how I was able to stream my desktop mic to my Quest 2 by selecting its ALSA device name. So streaming itself works fine.

When running alvr from the console I can see the following warnings, but I don't know if that's normal:

```
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
```

These are the ALSA audio devices using `aplay -L`:
```
null
    Discard all samples (playback) or generate zero samples (capture)
pipewire
    PipeWire Sound Server
default
    Default ALSA Output (currently PipeWire Media Server)
```

NixOS system info:
```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.68, NixOS, 23.11 (Tapir), 23.11.20231220.d65bcea`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/store/lnbafzxw52r0w55hwbcrsdkvk5fymv3l-source`
```

**Comment by NovaViper** [comment]
@GHOSCHT Hey did you download the[ audio setup script](https://github.com/alvr-org/ALVR-Distrobox-Linux-Guide/blob/main/audio-setup.sh)? The script is what I'm using to setup the virtual channels that ALVR streams to the headset. Looking at your nix config, it looks pretty much the same as my setup in regards of pipewire's setup. I also have [these lines in my ALVR config](https://github.com/NovaViper/NixConfig/blob/1968b5494e442f05148c53390c516c2551ce0f9b/home/novaviper/dots/alvr/session.json#L1012-L1013), which when I setup ALVR when I switched to the packaged version from the nightly appimage I had, I made ALVR reset my settings and went through the setup wizard.

**Comment by GHOSCHT** [comment]
> @GHOSCHT Hey did you download the[ audio setup script](https://github.com/alvr-org/ALVR-Distrobox-Linux-Guide/blob/main/audio-setup.sh)? The script is what I'm using to setup the virtual channels that ALVR streams to the headset. Looking at your nix config, it looks pretty much the same as my setup in regards of [pipewire's setup](https://github.com/NovaViper/NixConfig/blob/main/hosts/common/optional/desktop/pipewire.nix). I also have [these lines in my ALVR config](https://github.com/NovaViper/NixConfig/blob/1968b5494e442f05148c53390c516c2551ce0f9b/home/novaviper/dots/alvr/session.json#L1012-L1013), which when I setup ALVR when I switched to the packaged version from the nightly appimage I had, I made ALVR reset my settings and went through the setup wizard.

Thanks for the reminder. I actually did use and look at that script before posting, I realize now I should've included this part.
That's actually what actually made me think ALVR was supposed to spawn an additional pipewire node.
```sh
pactl move-source-output "$(get_playback_source_output_id alsa_capture.vrserver)" "$(get_sink_id_by_name ALVR-AUDIO-Sink)"
```
That command from the audio setup failed at `get_playback_source_output_id alsa_capture.vrserver` and I thought vrserver was a process from ALVR... Well after investigating a bit further I realized it was from SteamVR which means it was all because of SteamVR weirdness. I did a fresh reinstall of my system, fought a bit wih SteamVR error 109 (solved by a new alvr nightly fix) but finally got it to work with one of the newest ALVR Nightly versions. Tbh I can't really say I did anything major to fix it, just reinstall and wait for everything to be resolved.

TLDR: SteamVR was acting up and I thought it was thought by ALVR. Everything works now!

**Comment by PassiveLemon** [comment]
I fudged something up so please follow this PR instead: https://github.com/NixOS/nixpkgs/pull/284154

**Comment by PassiveLemon** [comment]
I've made a flake repository here: https://github.com/PassiveLemon/lemonake

You can find the ALVR module there and instructions to import it. I figured this is more convenient than manually adding the files to your config.

**Comment by PassiveLemon** [comment]
It's been merged. You can close this as solved now

**Comment by baracoder** [comment]
:tada:  A package was added in https://github.com/NixOS/nixpkgs/pull/284154 I think we can close this as completed
